server/models/Document.ts

```diff
@@ -436,32 +440,53 @@ class Document extends ParanoidModel {
     query: string,
     options: SearchOptions = {}
   ): Promise<SearchResponse> {
-    const limit = options.limit || 15;
-    const offset = options.offset || 0;
     const wildcardQuery = `${escape(query)}:*`;
-    const collectionIds = await team.collectionIds();
+    const { minWords = 20, maxWords = 30, limit = 15, offset = 0 } = options;
 
-    // If the team has access no public collections then shortcircuit the rest of this
+    // restrict to specific collection if provided
+    // enables search in private collections if specified
+    let collectionIds;
+    if (options.collectionId) {
+      collectionIds = [options.collectionId];
+    } else {
+      collectionIds = await team.collectionIds();
+    }
+
+    // short circuit if no relevant collections
     if (!collectionIds.length) {
       return {
         results: [],
         totalCount: 0,
       };
     }
 
-    // Build the SQL query to get documentIds, ranking, and search term context
+    // restrict to documents in the tree of a shared document when one is provided
+    let documentIds;
+
+    if (options.share?.includeChildDocuments) {
+      const sharedDocument = await options.share.$get("document");
+      if (sharedDocument) {
```

Comment:
This seems like a potential way to leak data, if the share document is not able to load for some reason, eg it's deleted then this `if` clause isn't hit and no restricting `documentIds` are added to the query.

Probably just throw here if `sharedDocument` not found?

Reply:
Fixed in the few lines below. I tried initially to do

```
const sharedDocument = await ...
invariant(sharedDocument, "some error");

// but TS thinks sharedDocument can still be null here
const childDocumentIds = await sharedDocument.getChildDocumentIds();
```

am I misunderstanding how invariant works or is it just temperamental?

Reply:
That should work, are you sure you imported the right invariant from the `invariant` package? I've noticed that sometimes VSCode will pull in a different one from elsewhere

Reply:
yeah that was the issue, changing it to that style now